File: src/main/java/com/nir/beans/Beans.java
package com.nir.beans;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.module.kotlin.KotlinModule;
import com.google.common.collect.Lists;
import com.nir.utils.math.method.generalized.GeneralizedMethodInfoJsonPojo;
import com.nir.utils.math.method.jsonPojos.AdamsBashforthGeneralizedMethodJsonPojo;
import com.nir.utils.math.method.jsonPojos.AdamsMoultonGeneralizedMethodJsonPojo;
import com.nir.utils.math.method.jsonPojos.ExplicitRKGeneralizedMethodJsonPojo;
import org.apache.commons.lang3.tuple.Pair;

import java.util.List;
import java.util.stream.Collectors;

public final class Beans {

    public static PeriodicElementsGetter periodicElementsGetter() {
        ObjectMapper kotlinMapper = kotlinObjectMapper();
        return new PeriodicElementsGetterImpl(kotlinMapper);
    }

    public static ObjectMapper kotlinObjectMapper() {
        ObjectMapper mapper = new ObjectMapper();
        mapper.registerModule(new KotlinModule());
//        mapper.setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE);
        mapper.configure(JsonParser.Feature.ALLOW_UNQUOTED_CONTROL_CHARS, true);
        mapper.enable(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY);
        return mapper;
    }

    public static GeneralizedMethodInfoComponentImpl generalizedMethodInfoComponent() {
        final List<JacksonComponent<GeneralizedMethodInfoJsonPojo>> jacksonComponents =
            Lists.newArrayList(
                Pair.of("json/runge-kutta.json", ExplicitRKGeneralizedMethodJsonPojo.class),
                Pair.of("json/adams-bashforth.json", AdamsBashforthGeneralizedMethodJsonPojo.class),
                Pair.of("json/adams-multon.json", AdamsMoultonGeneralizedMethodJsonPojo.class)
            )
            .stream()
            .map(Beans::jacksonComponent)
            .collect(Collectors.toList());
        return new GeneralizedMethodInfoComponentImpl(jacksonComponents);
    }

    public static JacksonComponent<GeneralizedMethodInfoJsonPojo> jacksonComponent(
        Pair<
            String,
            ? extends Class<? extends GeneralizedMethodInfoJsonPojo>
        >
            jsonAndClass
    )
    {
        final String filename = jsonAndClass.getKey();
        final Class<? extends GeneralizedMethodInfoJsonPojo> aClass = jsonAndClass.getValue();
        return new JacksonComponent<>(filename, aClass);
    }

    public static GeneralizedMethodComponent generalizedMethodComponent() {
        return new GeneralizedMethodComponent();
    }

    public static HardcodedMethodComponent hardcodedMethodComponent() {
        return new HardcodedMethodComponent();
    }
}

File: src/main/kotlin/com/nir/beans/Beans.kt
package com.nir.beans

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.KotlinModule
import com.nir.beans.method.MethodConverterComponent
import com.nir.beans.method.MethodInfoComponentImpl
import com.nir.beans.method.HardcodedMethodComponent
import com.nir.beans.method.MethodInfoComponent
import com.nir.beans.reaction.parsing.JacksonComponent
import com.nir.beans.method.generalized.MethodInfoJsonPojo
import com.nir.beans.method.jsonPojos.AdamsBashforthMethodJsonPojo
import com.nir.beans.method.jsonPojos.AdamsMoultonMethodJsonPojo
import com.nir.beans.method.jsonPojos.ExplicitRKMethodJsonPojo
import org.apache.commons.lang3.tuple.Pair

object Beans {
    @JvmStatic
    fun periodicElementsGetter(): PeriodicElementsGetter {
        val kotlinMapper: ObjectMapper = kotlinObjectMapper()
        return PeriodicElementsGetterImpl(kotlinMapper)
    }

    fun kotlinObjectMapper(): ObjectMapper {
        val mapper = ObjectMapper()
        mapper.registerModule(KotlinModule())
        //        mapper.setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE);
        mapper.configure(JsonParser.Feature.ALLOW_UNQUOTED_CONTROL_CHARS, true)
        mapper.enable(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY)
        return mapper
    }

    @JvmStatic
    fun methodInfoComponent(): MethodInfoComponent {
        val jacksonComponents: List<JacksonComponent<out MethodInfoJsonPojo>> = arrayListOf(
                Pair.of("json/runge-kutta.json", ExplicitRKMethodJsonPojo::class.java),
                Pair.of("json/adams-bashforth.json", AdamsBashforthMethodJsonPojo::class.java),
                Pair.of("json/adams-multon.json", AdamsMoultonMethodJsonPojo::class.java)
        )
                .map(Beans::jacksonComponent)
                .toList()

        return MethodInfoComponentImpl(jacksonComponents)
    }

    fun jacksonComponent(
            jsonAndClass: Pair<String, out Class<out MethodInfoJsonPojo>>
    ): JacksonComponent<out MethodInfoJsonPojo> {
        val filename = jsonAndClass.key
        val aClass: Class<out MethodInfoJsonPojo> = jsonAndClass.value
        return JacksonComponent(filename, aClass)
    }

    @JvmStatic
    fun methodConverterComponent(): MethodConverterComponent {
        return MethodConverterComponent()
    }

    @JvmStatic
    fun hardcodedMethodComponent(): HardcodedMethodComponent {
        return HardcodedMethodComponent()
    }
}